### DISCUSSION TITLE
[@types/lodash] Why pick has a partial overwrite?

### DISCUSSION THREAD

**Discussion by cangSDARM** [DefinitelyTyped/DefinitelyTyped#63009] [Issues with a @types package]
Two same logic, same object, different types:

First
![image](https://user-images.githubusercontent.com/20441896/199167838-6ac9ef2b-0f78-4268-b777-181dd031f7ab.png)
![image](https://user-images.githubusercontent.com/20441896/199167790-f8b97d18-b66d-4b86-b7c3-f3c3376daa93.png)

Second
![image](https://user-images.githubusercontent.com/20441896/199167813-a7e2f0cd-bd1e-450a-abe8-13a43bbbe982.png)
![image](https://user-images.githubusercontent.com/20441896/199167863-75ebcced-58b9-4261-ab83-ec23f1d0bc1b.png)

what happend??

[a comment by an automated account is not shown]

**Comment by cangSDARM** [comment]
I found the problem. When `pick<T, U extends keyof typeof T>`, and if U is not extends `keyof typeof T`, partial overwrite will take place. But using partial is wrong, or at least lazy. At this point the type should be `Pick<T, U> & { [newKey in Exclude<U, keyof typeof T>]: unknown | never }`

**Reply by aj-r** [reply]
In general you shouldn't pass properties to `pick` that don't exist on that type (for the same reason that TS throws an error if you write `FilterFormCommon.listingPlace`). IMO it should be an error to do that, but I believe that overload exists to avoid breaking code written before we switched to `Pick<T, U>`.

Can you change your code to ensure that `listingPlace` exists on `typeof FilterFormCommon`? Or remove `listingPlace` from the pick?

**Reply by cangSDARM** [reply]
Yes I Can. But if all such problems are solved by modifying the code, then TS is completely pointless, isn't it. In this case, the use of TS is to solve the "if listingPlace does not exist, then what do we do", not simply ignore/avoid the error

**Reply by aj-r** [reply]
Not exactly - the point of TS is to help you avoid mistakes. `pick(FilterFormCommon, 'listingPlace')` is almost certainly a mistake if we know that `listingPlace` doesn't exist at compile time. So we shouldn't support that (just like TS doesn't allow you to write `FilterFormCommon.listingPlace`). Arguably we should give a better error message in this case, but I'm not sure if that's feasible without breaking existing code.

**Reply by cangSDARM** [reply]
Yes, all I'm saying is to give a better error message in this case. `Partial` clearly defies it
